### PR TITLE
fix #2604 completely

### LINKED ISSUE #2604: The query entity throws an exception when a boolean property exists in the entity
## Bug Report

**For English only**, other languages will not accept.

Before report a bug, make sure you have:

- Searched open and closed [GitHub issues](https://github.com/apache/incubator-shardingsphere-example/issues).
- Read documentation: [ShardingSphere Doc](https://shardingsphere.apache.org/document/current/en/overview).

Please pay attention on issues you submitted, because we maybe need more details. 
If no response **more than 7 days** and we cannot reproduce it on current information, we will **close it**.

Please answer these questions before submitting your issue. Thanks!

### Which version of ShardingSphere did you use?
4.0.0-RC2-SNAPSHOT，4.0.0-RC1 version work well
### Which project did you use? Sharding-JDBC or Sharding-Proxy?
Sharding-JDBC
### Expected behavior
Work well
### Actual behavior
throw java.lang.ClassCastException: [B cannot be cast to java.lang.Boolean
### Reason analyze (If you can)
MySQL column type is BIT, MergeResultSet value is an byte array, can not convert to boolean.
### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.
pom.xml
```xml
        <properties>
		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
		<maven.compiler.source>1.8</maven.compiler.source>
		<maven.compiler.target>1.8</maven.compiler.target>
		<java.version>1.8</java.version>
		<!-- <sharding-sphere.version>4.0.0-RC1</sharding-sphere.version> -->
		<sharding-sphere.version>4.0.0-RC2-SNAPSHOT</sharding-sphere.version>
	</properties>
	
	<parent>
		<groupId>org.springframework.boot</groupId>
		<artifactId>spring-boot-starter-parent</artifactId>
		<version>2.0.1.RELEASE</version>
		<relativePath />
	</parent>

	<dependencies>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-web</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-data-jpa</artifactId>
		</dependency>
		<dependency>
			<groupId>mysql</groupId>
			<artifactId>mysql-connector-java</artifactId>
		</dependency>
		<dependency>
			<groupId>org.apache.shardingsphere</groupId>
			<artifactId>sharding-jdbc-spring-boot-starter</artifactId>
			<version>${sharding-sphere.version}</version>
		</dependency>
		<dependency>
			<groupId>org.projectlombok</groupId>
			<artifactId>lombok</artifactId>
			<scope>provided</scope>
		</dependency>
	</dependencies>
```
application.yml
```yml
spring:
  jpa:
    show-sql: true
    generate-ddl: true
  shardingsphere:
    datasource:
      names: d0,d1
      d0:
        type: com.zaxxer.hikari.HikariDataSource
        driver-class-name: com.mysql.jdbc.Driver
        jdbc-url: jdbc:mysql://dbserver:3306/d0?useUnicode=true&characterEncoding=utf-8&zeroDateTimeBehavior=convertToNull&useSSL=false
        username: root
        password:
      d1:
        type: com.zaxxer.hikari.HikariDataSource
        driver-class-name: com.mysql.jdbc.Driver
        jdbc-url: jdbc:mysql://dbserver:3306/d1?useUnicode=true&characterEncoding=utf-8&zeroDateTimeBehavior=convertToNull&useSSL=false
        username: root
        password:
    props:
      sql:
        show: true
    sharding:
      default-data-source-name: d0
      tables:
        demo:
          actual-data-nodes: d$->{0..1}.demo
          database-strategy:
            inline:
              sharding-column: id
              algorithm-expression: d$->{id % 2}
          key-generator:
            column: id
            type: SNOWFLAKE
```
Demo.java
```java
@Entity
@Table(name="demo")
@Setter
@Getter
@ToString
public class Demo {

	@Id
	@GeneratedValue(strategy=GenerationType.IDENTITY)
	private Long id;
	
	@Column
	private boolean test;
	
}
```
DemoRepository.java
```java
@Repository
public interface DemoRepository extends JpaRepository<Demo, Long> {

}
```
DemoApplication.java
```java
@SpringBootApplication
public class DemoApplication {
	
	public static void main(String[] args) {
		try (ConfigurableApplicationContext applicationContext = SpringApplication.run(DemoApplication.class, args)) {
			DemoRepository demoRepository = applicationContext.getBean(DemoRepository.class);
			Demo demo = new Demo();
			demo.setTest(true);
			demoRepository.saveAndFlush(demo);
			demoRepository.findById(demo.getId());
			demoRepository.deleteAll();
		}
	}
	
}
```
Log
```
Exception in thread "main" org.springframework.dao.InvalidDataAccessApiUsageException: [B cannot be cast to java.lang.Boolean; nested exception is java.lang.IllegalArgumentException: [B cannot be cast to java.lang.Boolean
	at org.springframework.orm.jpa.EntityManagerFactoryUtils.convertJpaAccessExceptionIfPossible(EntityManagerFactoryUtils.java:367)
	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:227)
	at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.translateExceptionIfPossible(AbstractEntityManagerFactoryBean.java:527)
	at org.springframework.dao.support.ChainedPersistenceExceptionTranslator.translateExceptionIfPossible(ChainedPersistenceExceptionTranslator.java:61)
	at org.springframework.dao.support.DataAccessUtils.translateIfNecessary(DataAccessUtils.java:242)
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:153)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.data.jpa.repository.support.CrudMethodMetadataPostProcessor$CrudMethodMetadataPopulatingMethodInterceptor.invoke(CrudMethodMetadataPostProcessor.java:135)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:92)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.data.repository.core.support.SurroundingTransactionDetectorMethodInterceptor.invoke(SurroundingTransactionDetectorMethodInterceptor.java:61)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
	at com.sun.proxy.$Proxy83.findById(Unknown Source)
	at net.saisimon.demo.DemoApplication.main(DemoApplication.java:19)
Caused by: java.lang.IllegalArgumentException: [B cannot be cast to java.lang.Boolean
	at org.hibernate.internal.SessionImpl.find(SessionImpl.java:3427)
	at org.hibernate.internal.SessionImpl.find(SessionImpl.java:3375)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.orm.jpa.SharedEntityManagerCreator$SharedEntityManagerInvocationHandler.invoke(SharedEntityManagerCreator.java:304)
	at com.sun.proxy.$Proxy81.find(Unknown Source)
	at org.springframework.data.jpa.repository.support.SimpleJpaRepository.findById(SimpleJpaRepository.java:235)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.data.repository.core.support.RepositoryComposition$RepositoryFragments.invoke(RepositoryComposition.java:377)
	at org.springframework.data.repository.core.support.RepositoryComposition.invoke(RepositoryComposition.java:200)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$ImplementationMethodExecutionInterceptor.invoke(RepositoryFactorySupport.java:629)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:593)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:578)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:59)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:294)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:98)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:139)
	... 10 more
Caused by: java.lang.ClassCastException: [B cannot be cast to java.lang.Boolean
	at org.apache.shardingsphere.shardingjdbc.jdbc.core.resultset.ShardingResultSet.getBoolean(ShardingResultSet.java:71)
	at org.hibernate.type.descriptor.sql.BooleanTypeDescriptor$2.doExtract(BooleanTypeDescriptor.java:59)
	at org.hibernate.type.descriptor.sql.BasicExtractor.extract(BasicExtractor.java:47)
	at org.hibernate.type.AbstractStandardBasicType.nullSafeGet(AbstractStandardBasicType.java:261)
	at org.hibernate.type.AbstractStandardBasicType.nullSafeGet(AbstractStandardBasicType.java:257)
	at org.hibernate.type.AbstractStandardBasicType.nullSafeGet(AbstractStandardBasicType.java:247)
	at org.hibernate.type.AbstractStandardBasicType.hydrate(AbstractStandardBasicType.java:333)
	at org.hibernate.persister.entity.AbstractEntityPersister.hydrate(AbstractEntityPersister.java:2854)
	at org.hibernate.loader.plan.exec.process.internal.EntityReferenceInitializerImpl.loadFromResultSet(EntityReferenceInitializerImpl.java:305)
	at org.hibernate.loader.plan.exec.process.internal.EntityReferenceInitializerImpl.hydrateEntityState(EntityReferenceInitializerImpl.java:233)
	at org.hibernate.loader.plan.exec.process.internal.AbstractRowReader.readRow(AbstractRowReader.java:103)
	at org.hibernate.loader.plan.exec.internal.EntityLoadQueryDetails$EntityLoaderRowReader.readRow(EntityLoadQueryDetails.java:241)
	at org.hibernate.loader.plan.exec.process.internal.ResultSetProcessorImpl.extractResults(ResultSetProcessorImpl.java:122)
	at org.hibernate.loader.plan.exec.internal.AbstractLoadPlanBasedLoader.executeLoad(AbstractLoadPlanBasedLoader.java:122)
	at org.hibernate.loader.plan.exec.internal.AbstractLoadPlanBasedLoader.executeLoad(AbstractLoadPlanBasedLoader.java:86)
	at org.hibernate.loader.entity.plan.AbstractLoadPlanBasedEntityLoader.load(AbstractLoadPlanBasedEntityLoader.java:167)
	at org.hibernate.persister.entity.AbstractEntityPersister.load(AbstractEntityPersister.java:4107)
	at org.hibernate.event.internal.DefaultLoadEventListener.loadFromDatasource(DefaultLoadEventListener.java:508)
	at org.hibernate.event.internal.DefaultLoadEventListener.doLoad(DefaultLoadEventListener.java:478)
	at org.hibernate.event.internal.DefaultLoadEventListener.load(DefaultLoadEventListener.java:219)
	at org.hibernate.event.internal.DefaultLoadEventListener.proxyOrLoad(DefaultLoadEventListener.java:278)
	at org.hibernate.event.internal.DefaultLoadEventListener.doOnLoad(DefaultLoadEventListener.java:121)
	at org.hibernate.event.internal.DefaultLoadEventListener.onLoad(DefaultLoadEventListener.java:89)
	at org.hibernate.internal.SessionImpl.fireLoad(SessionImpl.java:1239)
	at org.hibernate.internal.SessionImpl.access$1900(SessionImpl.java:203)
	at org.hibernate.internal.SessionImpl$IdentifierLoadAccessImpl.doLoad(SessionImpl.java:2803)
	at org.hibernate.internal.SessionImpl$IdentifierLoadAccessImpl.load(SessionImpl.java:2777)
	at org.hibernate.internal.SessionImpl.find(SessionImpl.java:3406)
	... 35 more
```

### PR DESCRIPTION
**Pull request by Saisimon**
Fixes #2604 .

Changes proposed in this pull request:
- use ResultSet.getObject() to handle BIT type, JDBC driver can return the correct result.
- handle MySQL BIT(n) to number case.
- add and modify some unit tests.

### CODE REVIEW

File: sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/core/execute/sql/execute/result/QueryResultUtil.java [hunk 1]
```diff
@@ -53,8 +53,6 @@ public static Object getValue(final ResultSet resultSet, final int columnIndex)
     public static Object getValueByColumnType(final ResultSet resultSet, final int columnIndex) throws SQLException {
         ResultSetMetaData metaData = resultSet.getMetaData();
         switch (metaData.getColumnType(columnIndex)) {
-            case Types.BIT:
```
Comment (sunbufu) [inline sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/core/execute/sql/execute/result/QueryResultUtil.java]: Why not just put `Types.BIT` behind `Types.BOOLEAN` ?

Comment (Saisimon) [reply sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/core/execute/sql/execute/result/QueryResultUtil.java]: You can find the answer from #2460 and #2467

Comment (tuohai666) [inline sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/core/execute/sql/execute/result/QueryResultUtil.java]: Please refer to #2460, the BIT was added to resolve a sharding-proxy issue. This change will make sharding-proxy return wrong value.

Comment (Saisimon) [reply sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/core/execute/sql/execute/result/QueryResultUtil.java]: > return field.isSingleBit() ? Boolean.valueOf(getBoolean(columnIndex)) : getBytes(columnIndex);

getObject() can return correct result

Comment (tuohai666) [reply sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/core/execute/sql/execute/result/QueryResultUtil.java]: Have you tested? Please show your test cases for sharding-proxy.

**Comment by Saisimon** [comment]
```java
@Entity
@Table(name="demo")
@Setter
@Getter
@ToString
public class Demo implements Serializable {
	
	private static final long serialVersionUID = -3946482552265049200L;

	@Id
	@GeneratedValue(strategy=GenerationType.IDENTITY)
	private Long id;
	
	@Column(columnDefinition="BIT(1)")
	private Boolean column1;
	
	@Column(columnDefinition="TINYINT(1)")
	private Boolean column2;
	
	@Column(columnDefinition="BIT(8)")
	private Byte column3;
	
	@Column(columnDefinition="BIT(16)")
	private Short column4;
	
	@Column(columnDefinition="BIT(32)")
	private Integer column5;
	
	@Column(columnDefinition="BIT(64)")
	private Long column6;
	
	@Column(columnDefinition="BIT(64)")
	private BigDecimal column7;
	
}
```

```java
@SpringBootApplication
public class DemoApplication {
	
	public static void main(String[] args) {
		try (ConfigurableApplicationContext applicationContext = SpringApplication.run(DemoApplication.class, args)) {
			DemoRepository demoRepository = applicationContext.getBean(DemoRepository.class);
			Demo demo = new Demo();
			demo.setColumn1(true);
			demo.setColumn2(true);
			demo.setColumn3(Byte.MAX_VALUE);
			demo.setColumn4(Short.MAX_VALUE);
			demo.setColumn5(Integer.MAX_VALUE);
			demo.setColumn6(Long.MAX_VALUE);
			demo.setColumn7(new BigDecimal(Long.MAX_VALUE));
			demoRepository.save(demo);
			
			demo = new Demo();
			demo.setColumn1(false);
			demo.setColumn2(false);
			demo.setColumn3((byte) 0);
			demo.setColumn4((short) 0);
			demo.setColumn5(0);
			demo.setColumn6(0L);
			demo.setColumn7(new BigDecimal(0));
			demoRepository.save(demo);
			
			List<Demo> demos = demoRepository.findAll();
			for (Demo d : demos) {
				System.out.println(d.toString());
			}
			
			demoRepository.deleteAll();
		}
	}
	
}
```
Tested.

**Comment by tuohai666** [comment]
Is it the test for sharding-proxy?

**Comment by Saisimon** [comment]
> Is it the test for sharding-proxy?

yes

File: sharding-proxy/sharding-proxy-transport/sharding-proxy-transport-mysql/src/main/java/org/apache/shardingsphere/shardingproxy/transport/mysql/packet/command/query/text/MySQLTextResultSetRowPacket.java [hunk 2]
```diff
@@ -64,6 +64,8 @@ public void write(final MySQLPacketPayload payload) {
                     payload.writeStringLenenc(each.toString().split("\\.")[0]);
                 } else if (each instanceof BigDecimal) {
                     payload.writeStringLenenc(((BigDecimal) each).toPlainString());
+                } else if (each instanceof Boolean) {
```
[the remark this answers is not in the document]
Comment (Saisimon) [reply sharding-proxy/sharding-proxy-transport/sharding-proxy-transport-mysql/src/main/java/org/apache/shardingsphere/shardingproxy/transport/mysql/packet/command/query/text/MySQLTextResultSetRowPacket.java]: When column type is BIT(1) or TINYINT(1), QueryResultUtil.getValueByColumnType() return boolean type.This is fine for Sharding-jdbc, but boolean type will be treated as a string when writing a packet in Sharding-proxy, the mysql driver in the user program receives a boolean value in the form of a string ("true" or "false"), the mysql driver directly converts it to a long. These two values are always greater than zero, so they return true each time. Sharding-proxy needs to return the correct byte array for the mysql driver in the user program to handle this case.

File: sharding-jdbc/sharding-jdbc-core/src/main/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/resultset/ResultSetUtil.java [hunk 3]
```diff
@@ -123,4 +131,20 @@ private static Object convertDateValue(final Object value, final Class<?> conver
...
+            case 1:
+                return convertNumberValue(bs[0], convertType);
+            case Shorts.BYTES:
+                return convertNumberValue(Shorts.fromByteArray(bs), convertType);
+            case Ints.BYTES:
+                return convertNumberValue(Ints.fromByteArray(bs), convertType);
+            case Longs.BYTES:
+                return convertNumberValue(Longs.fromByteArray(bs), convertType);
+            default:
+                return value;
```
[the remark this answers is not in the document]
Comment (Saisimon) [reply sharding-jdbc/sharding-jdbc-core/src/main/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/resultset/ResultSetUtil.java]: The length of the byte array in this place can only be 1, 2, 4, 8.

**Comment by tuohai666** [comment]
Accept. Thanks for your effort.

**Pull request merged**